includes/core/server/request/class-update-intention.php

```diff
@@ -0,0 +1,196 @@
+<?php
+/**
+ * Class file for WCPay\Core\Server\Request\Update_Intent.
+ *
+ * @package WooCommerce Payments
+ */
+
+namespace WCPay\Core\Server\Request;
+
+use WC_Payments;
+use WCPay\Core\Exceptions\Server\Request\Invalid_Request_Parameter_Exception;
+use WCPay\Core\Server\Request;
+use WC_Payments_API_Client;
+use WCPay\Payment_Methods\CC_Payment_Method;
+use WCPay\Payment_Methods\Link_Payment_Method;
+
+/**
+ * Request class for creating intents.
+ */
+class Update_Intention extends Request {
+	use Intention;
+	use Level3;
+
+	const IMMUTABLE_PARAMS = [ 'amount' ];
+	const REQUIRED_PARAMS  = [ 'amount', 'currency' ];
+	const DEFAULT_PARAMS   = [
+		'receipt_email' => '',
+		'metadata'      => [],
+	];
+
+	/**
+	 * Intent id.
+	 *
+	 * @var string|null $intent_id
+	 */
+	private $intent_id = null;
+
+	/**
+	 * Set intent id.
+	 *
+	 * @param string $intent_id Intent id.
+	 *
+	 * @return void
+	 * @throws Invalid_Request_Parameter_Exception
+	 */
+	public function set_intent_id( string $intent_id ) {
+		$this->validate_stripe_id( $intent_id, [ 'pi' ] );
+
+		// Prevent mutation of intent id. It can be only set once.
+		if ( null === $this->intent_id ) {
+			$this->intent_id = $intent_id;
```

Comment:
This would silently ignore the new ID if one is present, an exception would be better.

Why not just set it as an immutable param though?

Reply:
Immutable param is part of params property and this is part of endpoint. Maybe I can throw immutable exception here or stick to constructor injection to prevent confusion between params, setters and IDs in endpoints.